### ISSUE TITLE
Firebase v9 loads a large iframe.js file but only on mobile

### ISSUE THREAD

**Issue by ludwigbacklund** [firebase/firebase-js-sdk#4946]
### [REQUIRED] Describe your environment

  * Operating System version: macOS v11.3
  * Browser version: Chrome 90.0.4430.212
  * Firebase SDK version: firebase@9.0.0-beta.2
  * Firebase Product: auth

### [REQUIRED] Describe the problem

After upgrading from v8 to v9 a request for an iframe (`https://[projectname].firebaseapp.com/__/auth/iframe.js`) has started appearing on every page load, but only on mobile (via Chrome's simulated Device mode at least, and when Lighthouse is auditing the website).

An older Lighthouse report of our application, from before the upgrade to v9, did not mention this file at all so I can only assume it wasn't loaded back then.

This iframe file is big and seems unnecessary to our application since we never use any kind of iframing of Firebase functionality and the only auth login method we use is email login. 

Is there a way to disable the loading of this iframe?

![image](https://user-images.githubusercontent.com/13007299/119844422-b4450d80-bf08-11eb-8e1d-da2a71062544.png)

**Comment by looptheloop88** [comment]
Hi @ludwigbacklund, thanks for the report. This iframe from Firebase Auth is used for third party login and eagerly loaded on mobile in certain situations. Currently, there’s no way to disable it. You may refer to [this](https://github.com/firebase/firebase-js-sdk/blob/master/packages-exp/auth-exp/src/platform_browser/popup_redirect.ts#L168) for your reference.

**Comment by ludwigbacklund** [comment]
@looptheloop88 I see, thank you. However, I think you're a bit quick to close this. 263 kb of entirely unused code is a lot to download, parse and evaluate for nothing when our application has no use for it. I can imagine there are many other applications where this is true as well. For now we've had to use [patch-package](https://www.npmjs.com/package/patch-package) to manually disable the loading of this file which improved our Lighthouse score.

**Comment by looptheloop88** [comment]
Hi @ludwigbacklund, thanks for the feedback. We are exploring for potential solutions, however we can't provide definite timelines or details. I have reopened this issue and added a feature request label for the meantime.

**Comment by mark922** [comment]
@ludwigbacklund Can you share the snippet of your solution?

**Comment by ludwigbacklund** [comment]
@mark922 Open `node_modules/@firebase/auth/dist/esm2017/index-somehash.js`. Find the `_shouldInitProactively()` function. Change `return _isMobileBrowser() || _isSafari() || _isIOS();` to `return false`. Follow the patch-package README instructions to generate the patch and commit it.

**Comment by nikhilag** [comment]
Shouldn't this issue be given higher priority, since the benefit of firebase js package becoming smaller is nullified by this issue?

**Comment by sjbuysse** [comment]
@ludwigbacklund there are multiple index-somehash.js files, which one should I take? Or all of them? Why are there multiple? (they all look the same)

**Comment by ludwigbacklund** [comment]
> @ludwigbacklund there are multiple index-somehash.js files, which one should I take? Or all of them? Why are there multiple? (they all look the same)

Search in the @firebase/auth node_modules directory for `_shouldInitProactively` and you'll find it.

**Comment by nikhilag** [comment]
@looptheloop88 Could you please share if there's any update on this? If this issue is fixed, it will help us meet the core web vitals for an ecommerce site that we built for our client. As per pagespeed insights, if this is fixed, it will improve our LCP quite a bit (saw a very big improvement for desktop after moving to v9 but for mobile the score became worse).

**Comment by sam-gc** [comment]
Hi folks, the iframe code is in fact required if you're doing `signInWithPassword` or `signInWithRedirect`. If you're not using those methods (i.e. you're not using any of the OAuth providers), you can initialize Auth in a way that does not cause this code to load.

The default in [`getAuth()`](https://firebase.google.com/docs/reference/js/v9/auth.md#getauth) pulls in all dependencies you might need, including [`browserPopupRedirectResolver`](https://firebase.google.com/docs/reference/js/v9/auth#browserpopupredirectresolver). You can tailor which dependencies are pulled in by using [`initializeAuth`](https://firebase.google.com/docs/reference/js/v9/auth.md#initializeauth) instead of `getAuth`.

The notes in the reference doc for `initializeAuth` explain in more detail, but at a high level you can mimic the behavior of `getAuth()` without pulling in the iframe code by using this code instead:

```js
// initializeAuth throws an error if it's called more than once for the same app; save the reference.
const auth =initializeAuth(app, {
  persistence: [indexedDBLocalPersistence, browserLocalPersistence]
});
```

You can verify this by looking at the code: https://github.com/firebase/firebase-js-sdk/blob/6564e99ec16612960c7ea49a368fc18197a5f2d1/packages-exp/auth-exp/index.ts#L137-L140
As you can see, the snippet I wrote earlier just omits the `popupRedirectResolver` dependency. This will prevent the iframe code from loading. **But remember, this will prevent you from using `signInWithPopup` and `signInWithRedirect`.**

**Comment by ArnaudD** [comment]
Hi @sam-gc, thank you your response. Is it possible to delay the iframe code from loading until `signInWithPopup` or `signInWithRedirect` are called ?

**Comment by sam-gc** [comment]
In general it does delay. It only loads early in a few cases:
https://github.com/firebase/firebase-js-sdk/blob/7028c1159b9153f14c22a81fdd09df68ffcfcfb6/packages-exp/auth-exp/src/platform_browser/popup_redirect.ts#L166-L169

**Comment by nikhilag** [comment]
@sam-gc Is there no way to fix this such that lighthouse score doesn't penalize performance score for using firebase sdk?

**Comment by ivanvanderbyl** [comment]
Any updates on this? We'd definitely like to disable it on mobile until it's _really_ needed.

**Comment by sam-gc** [comment]
Hi folks, it is indeed possible to delay loading that code until the sign in methods are called. The popup and redirect family of methods take an optional third parameter that is the `popupRedirectResolver`. For example, the signature for [`signInWithPopup()`](https://firebase.google.com/docs/reference/js/auth#signinwithpopup) is as follows:

```ts
export declare function signInWithPopup(auth: Auth, provider: AuthProvider, resolver?: PopupRedirectResolver): Promise<UserCredential>;
```

So instead of using `getAuth()` (or `initializeAuth` with the `popupRedirectResolver` set---these are the same), you can call `initializeAuth()` **without** the `popupRedirectResolver` option, then pass it in to the sign in functions as you need them. For example, this code below will _always_ delay loading the iframe code until it is used:

```js
import {initializeAuth, indexedDBLocalPersistence, browserLocalPersistence, browserSessionPersistence, browserPopupRedirectResolver, signInWithPopup, GoogleAuthProvider} from 'firebase/auth';
const auth = initializeAuth(app, {
 persistence: [
    indexedDBLocalPersistence,
    browserLocalPersistence,
    browserSessionPersistence
  ],
});

async function signIn() {
  const result = await signInWithPopup(auth, new GoogleAuthProvider(), browserPopupRedirectResolver);
}
```

**Comment by mdathersajjad** [comment]
If anyone is looking for a patch. https://gist.github.com/mdathersajjad/628c53913c10f7e090d52871faf1c373. And add browserPopupRedirectResolver as given in above comment

**Comment by abdo643-HULK** [comment]
> Hi folks, it is indeed possible to delay loading that code until the sign in methods are called. The popup and redirect family of methods take an optional third parameter that is the `popupRedirectResolver`. For example, the signature for [`signInWithPopup()`](https://firebase.google.com/docs/reference/js/auth#signinwithpopup) is as follows:
> 
> ```ts
> export declare function signInWithPopup(auth: Auth, provider: AuthProvider, resolver?: PopupRedirectResolver): Promise<UserCredential>;
> ```
> 
> So instead of using `getAuth()` (or `initializeAuth` with the `popupRedirectResolver` set---these are the same), you can call `initializeAuth()` **without** the `popupRedirectResolver` option, then pass it in to the sign in functions as you need them. For example, this code below will _always_ delay loading the iframe code until it is used:
> 
> ```js
> import {initializeAuth, indexedDBLocalPersistence, browserLocalPersistence, browserSessionPersistence, browserPopupRedirectResolver, signInWithPopup, GoogleAuthProvider} from 'firebase/auth';
> const auth = initializeAuth(app, {
>  persistence: [
>     indexedDBLocalPersistence,
>     browserLocalPersistence,
>     browserSessionPersistence
>   ],
> });
> 
> async function signIn() {
>   const result = await signInWithPopup(auth, new GoogleAuthProvider(), browserPopupRedirectResolver);
> }
> ```

The only problem is that you'll get a pop-up blocked error on the first trying to login with a provider on Safari.

**Comment by sam-gc** [comment]
@abdo643-HULK yep, that's one of the reasons why the `_shouldInitProactively` functionality exists

**Comment by abdo643-HULK** [comment]
> @abdo643-HULK yep, that's one of the reasons why the `_shouldInitProactively` functionality exists

Where can I find this function ? I am looking through the source code of the auth but can't seem to find it. And thnks for your help

**Comment by sam-gc** [comment]
It's this code here: https://github.com/firebase/firebase-js-sdk/blob/49b0406abb9b211c5b75325b0383539ac03358d1/packages/auth/src/platform_browser/popup_redirect.ts#L166-L169

By omitting the `popupRedirectResolver` optional dependency in `initializeAuth()`, that function is never called

**Comment by Sdqumar** [comment]
Here goes my solution, i only dynamic [import](https://javascript.info/modules-dynamic-imports) firebase auth, when I need it.
So I don't call it in the entire application, only when am making request to firebase. 
If you have small to medium application, it should be to much work.
Tip* to keep then state of current user, use cookies 
![Screenshot_2021-09-28-06-02-55-097_com android chrome~2](https://user-images.githubusercontent.com/20935356/135026861-222aa6ca-e7e0-4cf9-98bf-98e83d467981.jpg)
![Screenshot_2021-09-28-06-02-42-884_com android chrome~2](https://user-images.githubusercontent.com/20935356/135026864-5594b304-11c9-486f-8fd6-5b8a599fd24e.jpg)

**Comment by vandres** [comment]
Using the examples from the comments, I get an error "TypeError: class constructors must be invoked with 'new'"



```ts
        provideAuth(() => {
            const auth: Auth = initializeAuth(getApp(), {
                persistence: [
                    indexedDBLocalPersistence,
                    browserLocalPersistence,
                    browserSessionPersistence
                ],
                popupRedirectResolver: undefined
            });
            if (environment.emulator) {
                connectAuthEmulator(auth, 'http://localhost:9099', {disableWarnings: true});
            }

            return auth;
        }),

 const credential: UserCredential = await signInWithPopup(this.auth, new GoogleAuthProvider(), browserPopupRedirectResolver);
```

**Comment by vandres** [comment]
Ah, the problem lies within AngularFire. Will open a ticket there

**Comment by Parakoos** [comment]
Another reason why this can be problematic is if you try to put the whole Auth module into a web worker. It works at first, until you test it on mobile, when suddenly it breaks because the web worker has no access to window. 

The use of initializeAuth instead of getAuth (as outlined by sam-gc) fixed the issue for me.

> ```js
> // initializeAuth throws an error if it's called more than once for the same app; save the reference.
> const auth =initializeAuth(app, {
>   persistence: [indexedDBLocalPersistence, browserLocalPersistence]
> });
> ```

PS. Having auth in a web worker already means that you cannot use signInWithPopup or Redirect, so that is not a drawback. To make that work for me using Google, I basically use Google Auth scripts to get a token, then send that token to my web worker auth script where I use signInWithCredential instead. :-)

**Comment by nasgnat** [comment]
I suspect this is what causes firebase to stop in PWA in offline mode. #5720

**Comment by kodai3** [comment]
https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-927800881

Is there nothing we could do to load iframe when we need ?

after initialization and before`signInWithPopup`

**Comment by buesing** [comment]
@sam-gc Your solution works great, except for on Safari. I have tried different hacks to get the script to load when I actually need it (on sign in/sign up pages) but couldn't get anything to work reliably. A method to initialize the loading of the script would be perfect, so we don't have to load it on every page.

**Comment by goellner** [comment]
Pagespeed Insights even says that it is not gzipped. This could shave off some loading time and could improve this issue while we wait for a proper solution.

**Comment by rejhgadellaa** [comment]
+1

If there's any way to get rid of the js, that would be sweet.

In the meantime, it should really get gzipped, it's a pretty bad hit on perf. The file is 263 kb unzipped while it could be ~83 kb (63% smaller). according to Pagespeed.

![image](https://user-images.githubusercontent.com/4703109/147411042-81ab7227-bba8-49cc-bf53-23501260b989.png)

**Comment by EricSteinberger** [comment]
Hi all, thanks for the work to fix this. Are there any updates on the timeline? This is likely hurting SEO and load times for many users out there and gzipping would fix 69% of the issue, so would be a high ROI change to make when you get around to it.

**Comment by TomBeckett** [comment]
It's strange that Google's own tool ([PageSpeed Insights](https://pagespeed.web.dev/)) recommends a solution (set a lifetime cache) for a Firebase which is extremely popular on mobile. Is there anyone internally you can pick this up and unblock?

There are various solutions above for lazy-loading Firebase Auth but it's not the recommended approach as per documentation since auth should ideally be initialised once and shared as much as possible.

It's sincerely been fantastic how far Firebase SDK has come to tackle bloat in the last year so this feels like an oversight?

**Comment by lewisd1996** [comment]
Desperate for a update on this, its killing my performance/lighthouse score

**Comment by sanny-io** [comment]
I was so shocked when I saw my measurements. This really needs a fix, haha.

**Comment by rejhgadellaa** [comment]
I've spun the compression issue into a separate ticket, see #6169. Now let's hope someone picks it up..

**Comment by diegoag64** [comment]
Any updates on this? I am finalizing a project ready to be put on production but its impossible to make the mobile pagespeed go above 70 with this iframe loading. I am still loading a 263kb package loaded on mobile every time..

**Comment by Parakoos** [comment]
> Any updates on this? I am finalizing a project ready to be put on production but its impossible to make the mobile pagespeed go above 70 with this iframe loading. I am still loading a 263kb package loaded on mobile every time..

You can do what I did and load all of firebase asynchronously into a web worker. So the initial load isn't affected at all by the size of firebase. Of course, you won't get any functionality until it is loaded but at least it is off the main thread and you can show whatever you need before the functionality is there.

**Comment by SanjivG10** [comment]
Need an update on this.

**Comment by webdev51** [comment]
Updates on this guy? Isn't this a very basic thing to do?

**Comment by pgdev108** [comment]
Firebase SDK Team,
This issue is a blocking for me to improve PageSpeed Insight. This issue has been open for sooooooooo long. 
Please assign this as urgent priority to fix it.

**Comment by jeremytenjo** [comment]
Hi @dwyfrequency 

Anyway you can help the firebase team prioritize this issue?

Thank you

**Comment by AsbarAli** [comment]
Any update on this issue?

**Comment by diegoag64** [comment]
Im also still waiting for an update on this

**Comment by kodai3** [comment]
@SanjivG10 

But that's not working well on Safari right ?

same as this 
https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-927800881

**Comment by dosstx** [comment]
Just wanted to say I am also dealing with this issue. Hoping for the best and will subscribe to this thread for updates!

**Comment by danielo515** [comment]
@SanjivG10 as @kodai3 mentioned, that will have several popup blocked issues, forcing the user to first, unblock the popup, second, re-try the authentication, which will probably generate a lot of support tickets in anyone's app support team.
I am testing the mentioned solution, and indeed the performance boost is HUGE, I went from 70 of performance to 97, and the time to first paint and to interactive were reduced significantly. Take a look at the screenshots:

![image](https://user-images.githubusercontent.com/2270425/177952366-e3a6b2bc-454e-41ab-9249-2e0a86eadbc3.png)
![image](https://user-images.githubusercontent.com/2270425/177952523-9754c70e-d544-4e27-8ac3-bae9c096a939.png)

The only difference between those two screenshots is deferring the load of the popup.
I wish I could just use this solution without any drawback, but seems that is not possible.
I will also happily isolate the loading of all auth modules (even with the pre-loaded popup) to the login page, but that is not possible either because you will need the auth module to check the user auth status AND the auth module can not be initialized twice, so you have to either pull everything into your application or nothing.

**Comment by Parakoos** [comment]
> The only difference between those two screenshots is deferring the load of the popup. I wish I could just use this solution without any drawback, but seems that is not possible. I will also happily isolate the loading of all auth modules (even with the pre-loaded popup) to the login page, but that is not possible either because you will need the auth module to check the user auth status AND the auth module can not be initialized twice, so you have to either pull everything into your application or nothing.

What I did was to isolate all firebase libraries in a Worker thread, that also loads the libraries lazily. Sure, now everything is asynchronous and like you said, the only way to check for user auth status is to load the library. So, everything that require user information is lazily loaded. I do, however, store some user information (very basic, like, are they logged in at all and what kind of access rights do I expect them to have) on local disk so that I can guess what menu options they will have. All that gets corrected once I get the real information, but this limits the redrawing since 99% of the time, the locally stored information is correct. 

The worker thread also has the added benefit of keeping the main thread more open, and it forced me to create an API between the threads. Does mean it will be easy to swap out Firebase if I ever need to! :-)

**Comment by rejhgadellaa** [comment]
@Parakoos' solution is basically what I do. I only load auth + auth-ui in the main thread if the user is not logged in.

Slightly offtopic, but Comlink (by dasurma) will save you a lot of work when moving firebase into a worker

**Comment by danielo515** [comment]
> What I did was to isolate all firebase libraries in a Worker thread, that also loads the libraries lazily. Sure, now everything is asynchronous and like you said, the only way to check for user auth status is to load the library. So, everything that require user information is lazily loaded. I do, however, store some user information (very basic, like, are they logged in at all and what kind of access rights do I expect them to have) on local disk so that I can guess what menu options they will have. All that gets corrected once I get the real information, but this limits the redrawing since 99% of the time, the locally stored information is correct.

@Parakoos, running all the firebase related operations has been in my mind for some years now. Not only because the size of le libraries (which is huge, specially in the pre-v9 era) but because all  of my projects are focused on being low cost but still need to compute statistics in slightly big datasets, so having all the async communication + big computations into a dedicated non blocking thread always looked very attractive to me. What was putting me away from it was that I was not sure about the investment (was it doable, or will I hit a roadblock after invested dozens of hours?) and because the complexity of the project which was already mature and big (not to mention the redux docs tell you to do all the calculations on the reducer but redux can not work in a worker... I should not have listen to them  😄 ). Now with your testimonial that is possible and starting a new and small project I feel more inclined to give t a go, and the existence of Comlink as pointed out by @rejhgadellaa really makes me want to try it out.

However, how do you manage the popup login scenario? I think I can maybe put all the auth logic into a lazy loaded component that, once the login is either verified or succeeded populates the app global store (or whatever you use) with the credentials and login information, and meanwhile just used whatever is cached in local storage? How are you sharing the credentials with the worker thread though? Do you send them from the main thread and re-initialize the firebase app there or how? I'm very interested in the topic

**Comment by rejhgadellaa** [comment]
You can find a small Comlink / firebase example in an issue I filed a while ago (since been resolved):
https://github.com/firebase/firebase-js-sdk/issues/5791

Regarding auth in the main thread; I currently do it in a Preact app, and have one route (/login) where I send the user when the auth in the worker signals the user isn't logged in, then I load everything I need in that route:

``` javascript
// Init firebase app
import { initializeApp } from 'firebase/app';
initializeApp({/* config  */ );

// Import firebase ui dynamically so we can do SSR
const firebaseUi = await import('firebaseui');

// Init & start Firebase Auth UI
const ui = new firebaseUi.auth.AuthUI( this._auth );
ui.start('#firebaseui-auth-container', /* config */);
```

The above is just some snippets of code I use, thrown together in a snippet, hope it helps :)

**Comment by Parakoos** [comment]
@danielo515 , I never load any firebase code in the main thread, so I don't use the AuthUI of course. Instead I use handmade login forms for email, google, facebook, twitter and github, get hold of the credentials, send them to the worker thread that uses them to log in to firebase. Basically, I use the 'Manual Flow' under all the Firebase Auth documentation. You can see this live at https://sharedgametimer.com/login

I did this because I didn't think I could do it any other way. I figured that logging the user in using AuthUI on the main thread wouldn't log the user in on the worker threat. But, seems @rejhgadellaa has gotten that to work (which would simplify the auth code A LOT). I do wonder if this means that the firebase core, app and auth modules are loaded twice? Once on the main thread, and again on the worker thread? I guess the download of the modules get cached (I'd hope) but parsing the JS has to happen twice, no? Is this a significant hit or a minor one? 

Perhaps this conversation is getting a bit off-topic. I'd love to continue the discussion off-thread, @rejhgadellaa  and @danielo515 . Email in my bio.

**Comment by MartinXPN** [comment]
Are there any updates on this issue? Would be really great to have faster performance for firebase on mobile.

**Comment by sam-gc** [comment]
Hi folks, the comment in https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-1163262324 is the "correct" solution. See also this commonet: https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-921147692.

Unfortunately, as you've noticed, this does not work in some cases like Safari. That is why this comment still applies: https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-884498397. It is not easy to fix this at all, because the iframe and popup code is necessary for the proper functioning of social sign in. We're still aware of this issue, but there are no immediate plans/designs in place to avoid proactive initialization in general.

**Comment by rejhgadellaa** [comment]
FWIW, the `iframe.js` does seem to have compression enabled ([for some time now](https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-1113449067)), but Lighthouse doesn't seem to pick it up and report it as uncompressed. So that's a false-positive.

I do wonder why I'm seeing 3 requests to `iframe.js` now (1x ~200b, 2x ~90kb, so one of them seems to be a dupe) but it's possible I have a bug in my code somewhere 😅

![image](https://user-images.githubusercontent.com/4703109/187247143-3150a2c2-2380-4e01-8422-80cec4bbb5ff.png)

**Comment by sundar-rajavelu** [comment]
Adding AngularFireAnalytics > UserTrackingService  in angular app > providers list loads the iframe immediately.

**Comment by raghavmri** [comment]
Hey, Is there any update on how to remove the load of iframe in next js?

**Comment by prameshj** [comment]
> Hey, Is there any update on how to remove the load of iframe in next js?

Hi @raghavmri , the iframe.js file is necessary if you use signInWithPopup or signInWithRedirect. 

Proactive loading of this file can be disabled by following the code snippet in https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-921147692, i see a next JS specific snippet in https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-1163262324

**Comment by MarioAda** [comment]
It's been more than 1.5 years. Will this ever be fixed?

**Comment by mhmo91** [comment]
This is exactly why Big tech companies shouldn't be able to legally buy startups... 2 years later, and such an massive performance hit still unresolved!

**Comment by meropis** [comment]
Some more information I'd like to put forward if it helps is that:

1. In order to help with debugging, this error persists for any browser that uses the webkit renderer. You can install a browser such as epiphany on linux where you can debug and test far more easily than loading up an IOS emulator etc;
2. If you are running a localhost development environment you'll find that when the error is thrown, if you have an existing auth object saved you can simply hot reload the environment (I use vite) and the auth will work correctly, even in an iframe;
3. CSP errors are fixable in most cases by using window.postMessage to make calls to external windows e.g. iframes. This is apparently not a possibility for this error after testing on my end. It is possible that someone with more experience may be able to find a way to pass the auth redirect back using postMessage and avoiding the CSP error, but I was unable to implement this.

I hope some of this will help in some way.

**Comment by atelog** [comment]
It is really bad that there is no simple solution to this problem. Why not simply give developers the option to dynamically import the module on demand?

As a temporary workaround in Angular, I conditionally provide the browserPopupRedirectResolver in the AppModule if the URL is the login page.
```ts
provideAuth(() => {
  if (typeof document !== 'undefined') {
    const isAuth = document.location.href.includes('auth');

    return initializeAuth(getApp(), {
      persistence: [
        indexedDBLocalPersistence,
        browserLocalPersistence,
        browserSessionPersistence,
      ],
      popupRedirectResolver: isAuth ? browserPopupRedirectResolver : undefined,
    });
  }
  return getAuth(getApp());
});
  ```
  
I also use the `href` attribute when navigating to the login page to reinitialize the Auth module. 
Additionally, I use `window.location.href` instead of the Angular router module in the guards.

This workaround improved the Lighthouse speed score, but it is super ugly.

**Comment by SamTech37** [comment]
Shouldn't this issue be fixed already?
It's taking forever!

**Comment by DmitryUlyanov** [comment]
Hi everyone, 

The solutions above only work if you don't have Google auth. Here is hack to make it work even with Google auth. It just overrides the function that decides whether to preload html or not.

```js
Object.defineProperty((browserPopupRedirectResolver as any).prototype, '_shouldInitProactively', {
    get: function () {
      return false;
    },
});
```

**Comment by ChrisChiasson** [comment]
I think we may be able to code our own workaround to separately call the `_initialize` method of `browserPopupRedirectResolver` at an arbitrary time if we could get a handle to an AuthInternal instance, which we might be able to find by inspecting auth dynamically in the console.

Specifically we would do something like the following--I'll report back what I find tomorrow:
const authInternal = someAuthInternalInstance
authInternal._popupRedirectResolver._initialize(authInternal)

Reference:
https://github.com/firebase/firebase-js-sdk/blob/4d88f0894b44b53875c295bd3af5044d04ff6e3a/packages/auth/src/platform_browser/popup_redirect.ts
https://github.com/firebase/firebase-js-sdk/blob/4d88f0894b44b53875c295bd3af5044d04ff6e3a/packages/auth/src/core/auth/auth_impl.ts#L160C1-L168C8

**Comment by ChrisChiasson** [comment]
Summary of investigation. Use the Object.defineProperty trick from Dimitry above to turn off the iframe, but then use this trick later to turn it on before you need to do the redirect. This avoids needing to reinitialize auth (the part danielo mentioned would be a blocker).

```
// here auth is the result of the standard getAuth, but prior to this you must have used Dimitry's Object.defineProperty trick
auth._popupRedirectResolver._initialize(auth);
```


Step by Step Details:
```
import { initializeApp } from "firebase/app";
import { browserPopupRedirectResolver, getAuth } from "firebase/auth";


const firebaseConfig = { /*...*/ };

function get() {
  console.trace("get _shouldInitProactively overridden"); // double check in the mobile simulator or tethered console
  return false;
}

Object.defineProperty(browserPopupRedirectResolver.prototype,
  '_shouldInitProactively', { get });

export const app = initializeApp(firebaseConfig);

const auth = getAuth();
console.log('auth._popupRedirectResolver before strigified / frozen', JSON.stringify(auth._popupRedirectResolver));


// if the next code line is never called, ctrl + f in the Elements dev tools tab will not find iframe.js.
// If it is called it will be found
auth._popupRedirectResolver._initialize(auth); // this is the key line that can initialize the iframe later instead of on startup
console.log('auth._popupRedirectResolver after (wait for console to update)', auth._popupRedirectResolver);
```

**Comment by SpectorHacked** [comment]
Its funny that its been almost 3 years, and still no fix, I am moving away to mongoDB just because of this issue.

**Comment by ChrisChiasson** [comment]
For others coming from search, [the post above](https://github.com/firebase/firebase-js-sdk/issues/4946#issuecomment-1918193756) does [fix the problem](https://stackoverflow.com/a/77938436). Granted, it wasn't fixed in Firebase Auth by Google directly, but the problem of unconditionally loading the `iframe.js` on mobile startup is reduced to practice now using the setup in that post. You now have the option of loading it only when the login with redirect or popup is about to happen. It doesn't produce the errors discussed in prior solutions, and does not require re-initializing Auth or Firebase itself.

**Comment by wahibimoh** [comment]
the following hack worked for me, I used replace-in-file npm package to patch the built bundle js using the following script:

```
import replace from 'replace-in-file'
const options = {
  files: 'dist/assets/index*.js',
  from: /get\s*_shouldInitProactively\(\)\s*\{[^\}]*/g,
  to: 'get _shouldInitProactively() { return false',
};

const results = await replace(options)
console.log('Replacement results:', results);
```

I made it run automatically postbuild, it works with minified an unminified builds. I hope it helps someone.

**Comment by satellite-xyz** [comment]
Experiencing the same issue here, in particular within my next.js app. The initial loading time is way too slow, and it seems mainly attributed to this.

**Comment by Tygovanommen** [comment]
+1 still an issue
